coderd/insights_test.go

```diff
@@ -570,6 +580,810 @@ func TestTemplateInsights(t *testing.T) {
+						},
+					},
+				},
+			},
+		}
+
+		return templates, users
+	}
+
+	// Time range for report, test data will be generated within and
+	// outside this range, but only data within the range should be
+	// included in the report.
+	frozenLastNight := time.Date(2023, 8, 22, 0, 0, 0, 0, time.UTC)
+	frozenWeekAgo := frozenLastNight.AddDate(0, 0, -7)
+
+	saoPaulo, err := time.LoadLocation("America/Sao_Paulo")
+	require.NoError(t, err)
+	frozenWeekAgoSaoPaulo, err := time.ParseInLocation(time.DateTime, frozenWeekAgo.Format(time.DateTime), saoPaulo)
+	require.NoError(t, err)
+
+	makeBaseTestData := func(templates []*testTemplate, users []*testUser) map[*testWorkspace]testDataGen {
+		return map[*testWorkspace]testDataGen{
+			users[0].workspaces[0]: {
+				agentStats: []agentStat{
+					{ // One hour of usage.
+						startedAt:          frozenWeekAgo,
+						endedAt:            frozenWeekAgo.Add(time.Hour),
+						sessionCountVSCode: 1,
+						sessionCountSSH:    1,
+					},
+					{ // 12 minutes of usage -> 15 minutes.
+						startedAt:       frozenWeekAgo.AddDate(0, 0, 1),
+						endedAt:         frozenWeekAgo.AddDate(0, 0, 1).Add(12 * time.Minute),
+						sessionCountSSH: 1,
+					},
+					{ // 2 minutes of usage -> 10 minutes because it crosses the 5 minute interval boundary.
+						startedAt:             frozenWeekAgo.AddDate(0, 0, 2).Add(4 * time.Minute),
+						endedAt:               frozenWeekAgo.AddDate(0, 0, 2).Add(6 * time.Minute),
+						sessionCountJetBrains: 1,
+					},
+				},
+				appUsage: []appUsage{
+					{ // One hour of usage.
+						app:       users[0].workspaces[0].apps[0],
+						startedAt: frozenWeekAgo,
+						endedAt:   frozenWeekAgo.Add(time.Hour),
+						requests:  1,
+					},
+					{ // used an app on the last day, counts as active user, 12m -> 15m rounded.
+						app:       users[0].workspaces[0].apps[2],
+						startedAt: frozenWeekAgo.AddDate(0, 0, 6),
+						endedAt:   frozenWeekAgo.AddDate(0, 0, 6).Add(12 * time.Minute),
+						requests:  1,
+					},
+				},
+			},
+			users[0].workspaces[1]: {
+				agentStats: []agentStat{
+					{
+						// One hour of usage in second template at the same time
+						// as in first template. When selecting both templates
+						// this user and their app usage will only be counted
+						// once but the template ID will show up in the data.
+						startedAt:          frozenWeekAgo,
+						endedAt:            frozenWeekAgo.Add(time.Hour),
+						sessionCountVSCode: 1,
+						sessionCountSSH:    1,
+					},
+				},
+				appUsage: []appUsage{
+					// TODO(mafredri): This doesn't behave correctly right now
+					// and will add more usage to the app. This could be
+					// considered both correct and incorrect behavior.
+					// { // One hour of usage, but same user and same template app, only count once.
+					// 	app:       users[0].workspaces[1].apps[0],
+					// 	startedAt: frozenWeekAgo,
+					// 	endedAt:   frozenWeekAgo.Add(time.Hour),
+					// 	requests:  1,
+					// },
```

Comment:
This is something I will fix later on, I wanted to keep query logic changes minimal for now.

Essentially this boils down to:
- User A has two workspaces in template X
- User A opens code-server in both workspaces
- User A closes both code-servers ater an hour
- We should only show 1h of code-server usage in template X

Currently we show two hours. Motivation for this is that a user can't effectively use multiple code-servers simultaneously. Imagine if this was 10 workspaces, the user would have to have very many skill points in parallel thinking and perhaps get a few extra keyboards and grow some extra limbs to produce 10h of usage. 😄